### PR TITLE
DownloadUpdateListener: to Kotlin

### PR DESCRIPTION
**Pull request by RingerJK**
## Description

Migrate Java to Kotlin  [#issue](https://github.com/mapbox/navigation-sdks/issues/167)

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    private val downloader: RouteTileDownloader,
+    private val tileUnpacker: TileUnpacker,
+    tilePath: String,
+    tileVersion: String,
+    private val listener: RouteTileDownloadListener
+) : DownloadTask.DownloadListener {
+
+    private val destinationPath: String
+
+    companion object {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt]: Should we extract this constant to a file constant and get rid of the `companion object`?

Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt]: Look suggestion above.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package com.mapbox.services.android.navigation.v5.navigation
+
+import com.mapbox.services.android.navigation.v5.utils.DownloadTask
+import java.io.File
+
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt]:
```suggestion

private const val DOWNLOAD_ERROR_MESSAGE =
    "Error occurred downloading tiles: null file found"
    
```

File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListenerTest.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import org.junit.Test
+import org.mockito.ArgumentMatchers.any
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.verify
+
+class DownloadUpdateListenerTest {
+
+    @Test
+    fun onFinishedDownloading_tarIsUnpacked() {
+        val tileUnpacker = mock<TileUnpacker>(TileUnpacker::class.java)
```
Comment (Guardiola31337) [inline libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListenerTest.kt]: We could remove explicit type arguments. The same thing happens in other variables below 👇

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt (hunk 1 above)
Comment (RingerJK) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt]: it's strong recommendation: for constant  use `companion object`. 
1. `const` inlines to code where it's used.
2. if we move `val DOWNLOAD_ERROR_MESSAGE` to class then it stays property and in java bytecode it looks like `filed` + `getField()` method

File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListenerTest.kt (hunk 3 above)
Comment (RingerJK) [reply libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListenerTest.kt]: agree, done

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt (hunk 1 above)
Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt]: > 1. const inlines to code where it's used.

Couldn't that be more performant? I also think the compiler does that behind the scenes with all constants.

> 2. if we move val DOWNLOAD_ERROR_MESSAGE to class then it stays property and in java bytecode it looks like filed + getField() method

Couldn't that be workarounded / 💅 using `@JvmStatic` annotations?

Comment (RingerJK) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/DownloadUpdateListener.kt]: 1. No, it's actually created property with 'getter' and in java it looks like 'getValue(): String'
2. @JvmStatic only for `companion object` and `Object` 😉

**Pull request merged**